File: backend/app/schemas/student.py
"""
Pydantic schemas for Student management.
"""

from typing import Optional, List
from datetime import datetime, date
from pydantic import BaseModel, EmailStr, Field, ConfigDict

from app.schemas.user import UserResponseSchema


class StudentCreate(BaseModel):
    """Schema for creating a new student (includes user data)."""

    # User fields
    email: EmailStr = Field(..., description="Student email address")
    password: str = Field(..., min_length=8, description="Student password (min 8 characters)")
    first_name: str = Field(..., min_length=1, max_length=100, description="First name")
    last_name: str = Field(..., min_length=1, max_length=100, description="Last name")
    middle_name: Optional[str] = Field(None, max_length=100, description="Middle name")
    phone: Optional[str] = Field(None, max_length=50, description="Phone number")

    # Student fields
    student_code: Optional[str] = Field(
        None,
        max_length=50,
        description="Student code (auto-generated if not provided)",
    )
    grade_level: int = Field(..., ge=1, le=11, description="Grade level (1-11)")
    birth_date: Optional[date] = Field(None, description="Date of birth")


class StudentUpdate(BaseModel):
    """Schema for updating an existing student."""

    # Only student-specific fields can be updated
    # User fields are updated via UserUpdate schema
    student_code: Optional[str] = Field(None, max_length=50, description="Student code")
    grade_level: Optional[int] = Field(None, ge=1, le=11, description="Grade level (1-11)")
    birth_date: Optional[date] = Field(None, description="Date of birth")


class SchoolClassBriefResponse(BaseModel):
    """Brief schema for class info (for nested use in Student response)."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    code: str
    grade_level: int
    academic_year: str


class StudentResponse(BaseModel):
    """Schema for student response (detailed with user data)."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    school_id: int
    user_id: int
    student_code: str
    grade_level: int
    birth_date: Optional[date]
    created_at: datetime
    updated_at: datetime

    # Nested user data
    user: UserResponseSchema

    # Nested classes data (optional, loaded when needed)
    classes: Optional[List[SchoolClassBriefResponse]] = None


class StudentListResponse(BaseModel):
    """Schema for student list response (brief)."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    student_code: str
    grade_level: int
    created_at: datetime

    # Nested user data (brief)
    user: UserResponseSchema

File: admin-v2/src/lib/validations/student.ts
import { z } from 'zod';

// Schema for creating a new student
export const studentCreateSchema = z.object({
  email: z
    .string()
    .min(1, 'Email обязателен')
    .email('Неверный формат email'),
  password: z
    .string()
    .min(6, 'Минимум 6 символов')
    .max(100, 'Максимум 100 символов'),
  first_name: z
    .string()
    .min(1, 'Имя обязательно')
    .max(100, 'Максимум 100 символов'),
  last_name: z
    .string()
    .min(1, 'Фамилия обязательна')
    .max(100, 'Максимум 100 символов'),
  middle_name: z
    .string()
    .max(100, 'Максимум 100 символов')
    .optional()
    .or(z.literal('')),
  phone: z
    .string()
    .max(50, 'Максимум 50 символов')
    .optional()
    .or(z.literal('')),
  student_code: z
    .string()
    .max(50, 'Максимум 50 символов')
    .optional()
    .or(z.literal('')),
  grade_level: z
    .number()
    .min(1, 'Минимум 1 класс')
    .max(11, 'Максимум 11 класс'),
  birth_date: z
    .string()
    .optional()
    .or(z.literal('')),
  enrollment_date: z
    .string()
    .optional()
    .or(z.literal('')),
});

// Schema for updating a student (without email/password)
export const studentUpdateSchema = z.object({
  first_name: z
    .string()
    .min(1, 'Имя обязательно')
    .max(100, 'Максимум 100 символов')
    .optional(),
  last_name: z
    .string()
    .min(1, 'Фамилия обязательна')
    .max(100, 'Максимум 100 символов')
    .optional(),
  middle_name: z
    .string()
    .max(100, 'Максимум 100 символов')
    .optional()
    .or(z.literal('')),
  phone: z
    .string()
    .max(50, 'Максимум 50 символов')
    .optional()
    .or(z.literal('')),
  student_code: z
    .string()
    .max(50, 'Максимум 50 символов')
    .optional()
    .or(z.literal('')),
  grade_level: z
    .number()
    .min(1, 'Минимум 1 класс')
    .max(11, 'Максимум 11 класс')
    .optional(),
  birth_date: z
    .string()
    .optional()
    .or(z.literal('')),
  enrollment_date: z
    .string()
    .optional()
    .or(z.literal('')),
});

// Infer types from schemas
export type StudentCreateInput = z.infer<typeof studentCreateSchema>;
export type StudentUpdateInput = z.infer<typeof studentUpdateSchema>;

// Default values for forms
export const studentCreateDefaults: StudentCreateInput = {
  email: '',
  password: '',
  first_name: '',
  last_name: '',
  middle_name: '',
  phone: '',
  student_code: '',
  grade_level: 7,
  birth_date: '',
  enrollment_date: '',
};
